From 153b041f182e0203a5c6653224eb1842b1a7d13d Mon Sep 17 00:00:00 2001
From: Jiang Lu <lu.jiang@windriver.com>
Date: Mon, 12 Oct 2015 09:50:10 +0000
Subject: [PATCH 216/216] dpaa2-eth:disable bottom halves in ldpaa_eth_open()

napi_alloc_frag should only be invoked in softirq context, disable
bottom halves in ldpaa_eth_open() to prevent following calltrace:

[ 5575.281618] BUG: using smp_processor_id() in preemptible [00000000] code: ifconfig/332
[ 5575.288240] caller is debug_smp_processor_id+0x1c/0x28
[ 5575.292095] CPU: 0 PID: 332 Comm: ifconfig Not tainted 4.1.6-WR8.0.0.0_standard #19
[ 5575.298450] Hardware name: Freescale Layerscape 2085a RDB Board (DT)
[ 5575.303509] Call trace:
[ 5575.304688] [<ffff8000000898c8>] dump_backtrace+0x0/0x128
[ 5575.308791] [<ffff800000089a10>] show_stack+0x20/0x30
[ 5575.312539] [<ffff80000079b78c>] dump_stack+0x80/0xc4
[ 5575.316543] [<ffff80000043e1ec>] check_preemption_disabled+0x10c/0x118
[ 5575.323166] [<ffff80000043e210>] debug_smp_processor_id+0x18/0x28
[ 5575.329344] [<ffff800000668650>] __alloc_page_frag+0x28/0x160
[ 5575.335158] [<ffff8000006687fc>] napi_alloc_frag+0x2c/0x40
[ 5575.340743] [<ffff800000645684>] ldpaa_bp_add_7.isra.4+0x4c/0x248
[ 5575.346902] [<ffff80000079d7c0>] ldpaa_eth_open+0x80/0x210
[ 5575.351529] [<ffff80000067e714>] __dev_open+0xcc/0x150
[ 5575.355361] [<ffff80000067ea40>] __dev_change_flags+0x90/0x160
[ 5575.359901] [<ffff80000067eb40>] dev_change_flags+0x30/0x70
[ 5575.364170] [<ffff8000006f1e08>] devinet_ioctl+0x680/0x728
[ 5575.368349] [<ffff8000006f3810>] inet_ioctl+0xe8/0x128
[ 5575.372194] [<ffff80000065ce34>] sock_ioctl+0x164/0x2f8
[ 5575.376116] [<ffff8000001deb10>] do_vfs_ioctl+0x358/0x5f0
[ 5575.380220] [<ffff8000001dee38>] SyS_ioctl+0x90/0xa8

Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
---
 drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
index 3cad0ab..fbd08c6 100644
--- a/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
+++ b/drivers/staging/fsl-dpaa2/ethernet/dpaa2-eth.c
@@ -1525,7 +1525,9 @@ static int ldpaa_dpbp_seed(struct ldpaa_eth_priv *priv, uint16_t bpid)
 
 	for_each_possible_cpu(j) {
 		for (i = 0; i < LDPAA_ETH_NUM_BUFS; i += 7) {
+			local_bh_disable();
 			new_count = ldpaa_bp_add_7(priv, bpid);
+			local_bh_enable();
 			count = per_cpu_ptr(priv->buf_count, j);
 			*count += new_count;
 
-- 
2.0.2

